### DISCUSSION TITLE
Automation of functionality in Slicer

### DISCUSSION THREAD

**Discussion by lukasvdstricht** [Project-MONAI/MONAILabel#1004] [Q&A]
Hi everyone

I am using MONAILabel together with Slicer to train models to automatically segment certain body parts on MRI/CT images.
The training went well and I have a decent model. I am interested in volumetric properties of the segmentation I have made with Automatic segmentation.
With the help of some existing blog posts and such, I have already succeeded in writing a Python script to generate a table with the SegmentStatistics module and save it locally as a .csv file. I would however want to automate this process even further and include in my script the step where I press Run for the automatic segmentation in MONAILabel. I am looking for how I can translate this step to Python code, similar to [this](https://discourse.slicer.org/t/using-segmentstatistics-module-in-python-script/2140).

I found the [source code](https://github.com/Project-MONAI/MONAILabel/blob/main/plugins/slicer/MONAILabel/MONAILabel.py) for this module, and I suspect that the function I’m looking for is onClickSegmentation on [line 1446](https://github.com/Project-MONAI/MONAILabel/blob/df9de94ff8b668d896e5aed902338811a0aca281/plugins/slicer/MONAILabel/MONAILabel.py#L1446). The only problem is that I don’t know how to invoke this function in a Python script.

Thanks in advance!

Lukas

**Comment by SachidanandAlle** [comment]
@tangy5 can you look into this question

**Comment by diazandr3s** [comment]
Hi @lukasvdstricht,

Do you want to use the exact code in Slicer? If you're running the MONAI Label server on the same PC, you could create a post-transform on the server side that preprocess the predicted label and store the output CSV file.

Here is the post transforms for the segmentation model that processed the model prediction: https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/radiology/lib/infers/segmentation.py#L69-L79

You may want to add a custom transform after the [AsDiscreted](https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/radiology/lib/infers/segmentation.py#L74) transform.

This is another idea. I'm sure @tangy5 can also help here :)

**Comment by tangy5** [comment]
HI @lukasvdstricht , thank you for raising the request and question. I think @diazandr3s mentioned above is a better choice, instead of changing the MONAI Label Plugin module source code, which required recompile afterward, or reinstall local MONAI Label package. The better option is to create your function/ script in the Sample-App directory here: https://github.com/Project-MONAI/MONAILabel/tree/main/sample-apps/radiology/lib, for example:

if you would like an automated metric calculation after inference/training, you could add function to the post-processing in /radiology/lib/infer/<your method script>.py or in /trainer, etc...

Also, it would be convenient and easy to add customize code to config for defining different models, inferer/trainer in the config folder such as spleen segmentation: https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/radiology/lib/configs/segmentation_spleen.py

Thank you, hope this helps.

**Reply by lukasvdstricht** [reply]
Hi @tangy5 

I don't think I'm trying to change the model source code. I just have some local `Python` files that I invoke in `Slicer`'s `Python Interactor` to do certain things automatically.
For example, this is the code that I have for saving  the table on the node called 'Table' to a `.csv` file:

```
def saveTable():
	tablenode = getNode('Table')
	slicer.util.saveNode(tablenode,"tableTest.csv")

saveTable()
```

The method offered by @diazandr3s  above may work for this particular task (creating and saving the table of volumetric properties) but in the future we may broaden our desire for automation to other tasks. So my question essentially is how you use `Slicer`'s `MONAILabel` module from a `Python` file.

I hope this makes it more clear.

Thank you very much for your answer!

Lukas
`

**Reply by pieper** [reply]
If you really need to, in Slicer's python interactor you can use code like this`slicer.modules.MONAILabelWidget.onClickSegmentation()` once the module widget has been constructed (use `slicer.util.selectModule("MONAILabel")` if you need to force this.  

However we try to avoid using the GUI methods for this kind of thing and instead try to use an API that doesn't mix GUI and Logic operations.  I haven't looked closely at the implementation, but ideally there would be a set of methods in MONAILabelLogic that mix Slicer functionality with MONAI functionality independent of the GUI, and that logic layer would leverage MONAI-specific functionality in the `MONAILabelClient`.

**Reply by lassoan** [reply]
Fully agree with @pieper. All the MONAILabel features that someone would want to run in a batch mode should be exposed in the logic class. This allows changing the GUI without worrying that it messes up things for people who use the module via Python scripting.

I've checked and there are already a number of methods in the logic class, which might be sufficient for you:

https://github.com/Project-MONAI/MONAILabel/blob/5a7cbd8d5de945a99085d0f6495de21485c47415/plugins/slicer/MONAILabel/MONAILabel.py#L2068-L2167

You can use it like this:

```python
logic = slicer.util.getModuleLogic('MONAILabel')
...
result_file, params = logic.infer(...)
```
